### ISSUE TITLE
Pythonize NetKet for a simple user experience

### ISSUE THREAD

**Issue by quantshah** [netket/netket#8]
Hi, thank you very much for releasing this tool. I am sure many in the community are looking forward to its use. 

I was wondering if it is possible to take full advantage of the Python ecosystem and develop this tool in a way similar to any Python-based scientific computing library. More specifically, the installation of the package can be made as simple as `pip install netket` or `conda install netket`. Python can handle running C++ code in the background and allows the use of [notebooks such as this one on superradiance](https://github.com/qutip/qutip-notebooks/blob/master/examples/piqs_superradiance.ipynb). It is a recent tool for dealing with permutational symmetry that we have developed and the package can be found in [PIQS](http://github.com/nathanshammah/piqs).

The current workflow where a JSON parameters file is created and then the C++ code is run using it can be completely done behind the scenes by wrapping everything up in a Python-based package. This will allow users to simply run your code in a python script or inside a Jupyter notebook. A simple prototype example could be:

```
from netket.graph import Hypercube
from netket.machine import RbmSpinSymm
from netket.hamiltonian import BoseHubbard
from netket import ground

graph = Hypercube(L=12, dim=1)
hamiltonian = BoseHubbard(4.0, nmax=3, bosons=12)
machine = RbmSpinSymm(alpha=4.0)

ground_state = ground(hamiltonian, graph, machine, method="Sr", learning_rate=0.1)
print(ground_state)
```

I am opening this issue in case you would like to develop this into a Python package.

**Comment by gcarleo** [comment]
Hi, and thanks a lot for those suggestions! 

Both things are certainly very interesting. 

The pip, conda etc installation strategies would be a nice add-on. Indeed the installation system I put in place right now is quite rudimentary, and should be improved. 

The only potential issue here is the fact that NetKet needs MPI libraries. I guess this would rule out pip install and leave only conda as a possibility. (I might be wrong!)  
Maybe for mac Os one could also write a simple brew module that would automatically resolve the dependence on MPI.

I have the feeling that the dependency on MPI can be also a bit problematic when writing the python interface you were proposing. But to be honest I have not looked too much into this, and It would be really nice if you had some prototype implementation you could contribute!

**Comment by sahmed95** [comment]
I think it is possible to specify compiler flags (mpicc) or write a custom build script for compiling the C/C++ code using mpicc upon installation using conda. You are right about pip though, it seems a hassle.

But tbh, this is majorly a C based package so it might just become an overstretch to try and write Python interfaces for everything. Unlike [QuTiP](qutip.org) or [PIQS](https://github.com/nathanshammah/piqs), I believe that the package was not written for Python first and then the computationally demanding parts optimized by C code. 

But maybe for users who are not running the code on a cluster or want a simpler interface an easy conda installation in a Python environment could be helpful. This also allows the use of Jupyter notebook examples. I hope I can stick around and help out. It is very kind of you to open-source your code. 

One might also take the gradient descent part of the code and wrap it up with Python functions. The rest of the code such as building a Hamiltonian, constructing a graph etc can be done using Python. It might even be easier that way for more users to play with the code, build arbitrary Hamiltonians, RBM architectures etc. as Python is easier to work with.

The beauty of your code is that you address the issue of complex numbers in a gradient descent method which is not supported directly in TensorFlow or PyTorch or other Machine Learning libraries. Let me get familiar with the codebase and I will try to see how a Python interface can be written for some of the part (say the gradient descent). But feel free to close the issue if you think this is an overstretch and a Python-based package is not the direction you are going for.

Thanks.

**Comment by gcarleo** [comment]
Sure, it is true that one might rewrite some of the less computationally intensive parts in python from scratch. That would certainly work for the Hamiltonian, Observables etc... I am not sure about the Machines though... for that part I believe it is quite critical to stick to C++ code! 

I am leaving this issue open and labeling it as "help wanted", maybe we'll manage to find a solution or others will want to give a hand

**Comment by gcarleo** [comment]
hi @sahmed95 , you might want to have a look at the branch "AllObjects". For example this commit: 588044aa420a9de313de06bdf567f0672a8972b5 . I streamlined the structure of the code, and I think it should be a easier now to port a few functions to python.
